datalad/config.py

```diff
@@ -71,44 +79,75 @@ def _where_reload(obj):
     return obj
 
 
-# TODO document and make "public" (used in GitRepo too)
-def _parse_gitconfig_dump(dump, cwd=None, multi_value=True):
+def parse_gitconfig_dump(dump, cwd=None, multi_value=True):
     """Parse a dump-string from `git config -z --list`
 
+    This parser has limited support for discarding unrelated output
+    that may contaminate the given dump. It does so performing a
+    relatively strict matching of configuration key syntax, and discarding
+    lines in the output that are not valid git-config keys.
+
+    There is also built-in support for parsing outputs generated
+    with --show-origin (see return value).
+
     Parameters
     ----------
     dump : str
       Null-byte separated output
     cwd : path-like, optional
-      Use this path to convert relative paths for origin reports
-      into absolute paths
+      Use this absolute path to convert relative paths for origin reports
+      into absolute paths. By default, the process working directory
+      PWD is used.
     multi_value : bool, optional
       If True, report values from multiple specifications of the
       same key as a tuple of values assigned to this key. Otherwise,
       the last configuration is reported.
+
+    Returns:
+    --------
+    dict, set
+      Configuration items are returned as key/value pairs in a dictionary.
+      The second tuple-item will be a set of path objects comprising all
+      source files, if origin information was included in the dump
+      (--show-origin). An empty set is returned otherwise.
     """
     dct = {}
     fileset = set()
     for line in dump.split('\0'):
-        if not line:
-            continue
-        if line.startswith('file:'):
-            # origin line
-            fname = Path(line[5:])
-            if not fname.is_absolute():
-                fname = Path(cwd) / fname if cwd else Path.cwd() / fname
-            fileset.add(fname)
-            continue
-        if line.startswith('command line:'):
-            # nothing we could handle
+        # line is a null-delimited chunk
+        k = None
+        # in anticipation of output contamination, process within a loop
+        # where we can reject non syntax compliant pieces
+        while line:
+            if line.startswith('file:'):
+                # origin line
+                fname = Path(line[5:])
+                if not fname.is_absolute():
+                    fname = Path(cwd) / fname if cwd else Path.cwd() / fname
+                fileset.add(fname)
+                break
+            if line.startswith('command line:'):
+                # no origin that we could as a pathobj
+                break
+            # try getting key/value pair from the present chunk
+            k, v = _gitcfg_rec_to_keyvalue(line)
+            if v or cfg_k_regex.match(k):
```

Comment:
Shouldn't the `cfg_k_regex.match(k)` happen within `_gitcfg_rec_to_keyvalue` instead?
I find it a bit weird to have the result of the latter be checked like that afterwards. Internally, that function already makes the assumption that it must be a key-only line, so just check it within ...

Reply:
The expression used in `_gitcfg_rec_to_keyvalue()` contains `cfg_k_regex`, so what you seem to ask for is already the case.

Reply:
Then there would be no need to test here again. See below.

Reply:
Ah, I get it now. Thx for the explanation below.